File: Prosary.Tests/DivineMercyEngineTests.cs
using Prosary.Models;
using Prosary.Services;
using Xunit;

namespace Prosary.Tests;

/// <summary>Mirrors iOS's DivineMercyEngineTests.swift / Android's DivineMercyEngineTest.kt.</summary>
public class DivineMercyEngineTests
{
    private readonly PrayerEngine _engine = new(new LiturgicalCalendarService());

    private static IReadOnlyList<RosaryStep> BuildSteps(PrayerEngine engine, string? languageCode) =>
        engine.BuildSteps(new Prayer { Kind = PrayerKind.DivineMercyChaplet, LanguageCode = languageCode ?? LanguageCatalog.DefaultSentinel });

    [Fact]
    public void BuildSteps_FiveDecadesOfTenPetitions()
    {
        var steps = BuildSteps(_engine, "en");
        var decadeIndices = steps.Where(s => s.DecadeIndex.HasValue).Select(s => s.DecadeIndex!.Value).Distinct();
        Assert.Equal(Enumerable.Range(0, 5), decadeIndices.OrderBy(i => i));

        for (var d = 0; d < 5; d++)
        {
            var petitionsInDecade = steps.Count(s => s.DecadeIndex == d && s.HailMaryIndexInDecade.HasValue);
            Assert.Equal(10, petitionsInDecade);
        }
    }

    [Fact]
    public void BuildSteps_NoStepHasAMystery()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.All(steps, s => Assert.Null(s.Mystery));
    }

    [Fact]
    public void BuildSteps_EveryStepReusesTheSingleDivineMercyImage()
    {
        // Unlike the Rosary/Franciscan Crown/Seven Sorrows, every step — opening, decades, and
        // closing alike — reuses the one divine_mercy_image illustration (the same reuse pattern
        // the Angelus uses for joyful_01_annunciation), since there's no per-decade content to
        // illustrate separately.
        var steps = BuildSteps(_engine, "en");
        Assert.All(steps, s => Assert.Equal("divine_mercy_image", s.ImageOverrideKey));
    }

    [Fact]
    public void BuildSteps_OfferingIsRepeatedIdenticallyAcrossEveryDecade()
    {
        var steps = BuildSteps(_engine, "en");
        var offerings = steps.Where(s => s.DecadeIndex.HasValue && !s.HailMaryIndexInDecade.HasValue).ToList();
        Assert.Equal(5, offerings.Count);
        Assert.All(offerings, s => Assert.Contains("Eternal Father, I offer You", s.Body));
    }

    [Fact]
    public void BuildSteps_OpeningReusesExistingPrayersNotNewContent()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.Equal("Sign of the Cross", steps[0].Title);
        Assert.Equal("Our Father", steps[1].Title);
        Assert.Equal("Hail Mary", steps[2].Title);
        Assert.Equal("The Apostles' Creed", steps[3].Title);
    }

    [Fact]
    public void BuildSteps_ClosingAcclamationRepeatedThreeTimesThenClosingCross()
    {
        var steps = BuildSteps(_engine, "en");
        var closingAcclamations = steps.Where(s => !s.DecadeIndex.HasValue && s.Body.Contains("Holy God")).ToList();
        Assert.Equal(3, closingAcclamations.Count);
        Assert.Equal("Sign of the Cross", steps[^1].Title);
    }

    [Fact]
    public void BuildSteps_EnglishBodyContainsEnglishText()
    {
        var steps = BuildSteps(_engine, "en");
        Assert.Contains(steps, s => s.Body.Contains("For the sake of His sorrowful Passion"));
    }

    [Fact]
    public void BuildSteps_LatinBodyContainsLatinText()
    {
        var steps = BuildSteps(_engine, "la");
        Assert.Contains(steps, s => s.Body.Contains("Pro dolorosa Eius passione"));
    }
}

File: iOS/ProsaryTests/DivineMercyEngineTests.swift
//
//  DivineMercyEngineTests.swift
//  ProsaryTests
//

import XCTest
@testable import Prosary

final class DivineMercyEngineTests: XCTestCase {
  private let engine = PrayerEngine()

  func testFiveDecadesOfTenPetitions() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    let decadeIndices = Set(steps.compactMap(\.decadeIndex))
    XCTAssertEqual(decadeIndices, Set(0..<5))

    for d in 0..<5 {
      let petitionsInDecade = steps.filter { $0.decadeIndex == d && $0.hailMaryIndexInDecade != nil }
      XCTAssertEqual(petitionsInDecade.count, 10, "decade \(d) should have 10 petitions")
    }
  }

  func testNoStepHasAMystery() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    XCTAssertTrue(steps.allSatisfy { $0.mystery == nil })
  }

  func testEveryStepReusesTheSingleDivineMercyImage() {
    // Unlike the Rosary/Franciscan Crown/Seven Sorrows, every step — opening, decades, and
    // closing alike — reuses the one divine_mercy_image illustration (the same reuse pattern
    // the Angelus uses for joyful_01_annunciation), since there's no per-decade content to
    // illustrate separately.
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    XCTAssertTrue(steps.allSatisfy { $0.imageKey == "divine_mercy_image" })
  }

  func testOfferingIsRepeatedIdenticallyAcrossEveryDecade() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    let offerings = steps.filter { $0.decadeIndex != nil && $0.hailMaryIndexInDecade == nil }
    XCTAssertEqual(offerings.count, 5)
    XCTAssertTrue(offerings.allSatisfy { $0.body.contains("Eternal Father, I offer You") })
  }

  func testOpeningReusesExistingPrayersNotNewContent() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    XCTAssertEqual(steps[0].title, "Sign of the Cross")
    XCTAssertEqual(steps[1].title, "Our Father")
    XCTAssertEqual(steps[2].title, "Hail Mary")
    XCTAssertEqual(steps[3].title, "The Apostles' Creed")
  }

  func testClosingAcclamationRepeatedThreeTimesThenClosingCross() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    let closingAcclamations = steps.filter { $0.decadeIndex == nil && $0.body.contains("Holy God") }
    XCTAssertEqual(closingAcclamations.count, 3)
    XCTAssertEqual(steps.last?.title, "Sign of the Cross")
  }

  func testEnglishBodyContainsEnglishText() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "en"))
    XCTAssertTrue(steps.contains { $0.body.contains("For the sake of His sorrowful Passion") })
  }

  func testLatinBodyContainsLatinText() {
    let steps = engine.buildSteps(for: Prayer(kind: .divineMercyChaplet, languageCode: "la"))
    XCTAssertTrue(steps.contains { $0.body.contains("Pro dolorosa Eius passione") })
  }
}
